File: src/service_bus/PortPool.cc
#include "PortPool.h"

#define LOG_LEVEL INFO_LEVEL
#include "Logger.h"

using namespace service_bus;

SharedQueue* PortPool::POOL = NULL;
unsigned int PortPool::PORT_LOWER = 0;
unsigned int PortPool::PORT_UPPER = 0;

void PortPool::initialize_statics(unsigned int port_lower, unsigned int port_upper) {
    if (port_lower > port_upper) {
        Utils::error("Invalid port limits [" + to_string(port_lower) + ".." + to_string(port_upper) +
                     "]");
    }
    LOG_INFO("Port range: [" << port_lower << " : " << port_upper << "]");
    PORT_LOWER = port_lower;
    PORT_UPPER = port_upper;
    POOL = new SharedQueue(port_upper - port_lower + 1);
    for (unsigned int port = port_lower; port <= port_upper; port++) {
        if (Utils::is_port_available(port)) {
            POOL->enqueue((void*) ((unsigned long) port));
        } else {
            LOG_ERROR("Discarding unavailable PORT " + std::to_string(port));
        }
    }
}

unsigned int PortPool::get_port() {
    unsigned int port = (unsigned int) ((unsigned long) POOL->dequeue());
    if (!port) {
        Utils::error("Unable to get available PORT number in [" + to_string(PORT_LOWER) + ".." +
                     to_string(PORT_UPPER) + "]");
    }
    return port;
}

void PortPool::return_port(unsigned int port) { POOL->enqueue((void*) ((unsigned long) port)); }

File: 3rd_party_slots/python_client/src/port_pool.py
from queue import Queue, Empty
from src.logger import log


class PortPool:
    _pool: Queue = None
    _port_lower: int = 0
    _port_upper: int = 0

    @classmethod
    def initialize_statics(cls, port_lower: int, port_upper: int) -> None:
        if port_lower > port_upper:
            raise ValueError(f"Invalid port limits [{port_lower}..{port_upper}]")
        log.info(f"Port range: [{port_lower} : {port_upper}]")
        cls._port_lower = port_lower
        cls._port_upper = port_upper
        cls._pool = Queue(maxsize=port_upper - port_lower + 1)
        for port in range(port_lower, port_upper + 1):
            cls._pool.put(port)

    @classmethod
    def get_port(cls) -> int:
        if cls._pool is None:
            raise ValueError("PortPool not initialized. Call initialize_statics first.")

        try:
            port = cls._pool.get(timeout=0.1)
        except Empty:
            raise ValueError(
                f"Unable to get available PORT number in [{cls._port_lower}..{cls._port_upper}]"
            )
        return port

    @classmethod
    def return_port(cls, port: int) -> None:
        if cls._pool is None:
            raise ValueError("PortPool not initialized. Call initialize_statics first.")

        if not (cls._port_lower <= port <= cls._port_upper):
            raise ValueError(f"Port out of the available range available [{cls._port_lower}..{cls._port_upper}]")

        cls._pool.put(port)
